service/history/shard/controller_impl.go

```diff
@@ -402,6 +414,12 @@ func (c *ControllerImpl) acquireShards(ctx context.Context) {
 
 	ctx = headers.SetCallerInfo(ctx, headers.SystemBackgroundCallerInfo)
 
+	// cancel previous readiness check goroutine
```

Comment:
I suggest making explicit what I think is happening here: it's intended that the readiness check does not run concurrently with the tryAcquire loop.

Reply:
Sort of.. my goal here wasn't really about concurrency, it's that the readiness check should be invalidated as soon as membership changes. I.e. fully acquiring one set of shards shouldn't count as "ready" if the set of owned shards is now different.